[popover] Fix popover-attribute-basic.html

rwlbuis opened the pull request:
#### e6f61f306afaa4ee95dc36040e166bfd037cfe75
<pre>
[popover] Fix popover-attribute-basic.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=253673">https://bugs.webkit.org/show_bug.cgi?id=253673</a>

Reviewed by Tim Nguyen.

In popoverAttributeChanged when hiding the popover make sure we fire beforetoggle event.
Since this event may end up changing the popover attribute again, recompute the type
after hiding the popover.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-change-type-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-change-type.html: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-change-type-expected.txt: Added.
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):
(WebCore::HTMLElement::popoverAttributeChanged):

Canonical link: <a href="https://commits.webkit.org/263188@main">https://commits.webkit.org/263188@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14b34f2c4087acf7505bef3377cc963012c33716

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3880 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3970 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5315 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4139 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3856 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4052 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/3975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3797 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3928 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4126 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3497 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5151 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1630 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3472 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5190 "9 flakes 136 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3449 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3532 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4922 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3934 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/3975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3455 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3472 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/949 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3500 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3734 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1373,18 +1373,22 @@ ExceptionOr<void> HTMLElement::showPopover()
 }
 
-ExceptionOr<void> HTMLElement::hidePopoverInternal(FocusPreviousElement focusPreviousElement, FireEvents fireEvents)
+ExceptionOr<void> HTMLElement::hidePopoverInternal(FocusPreviousElement focusPreviousElement, FireEvents fireEvents, SuppressExceptions suppressExceptions)
```
I'm not convinced we need this `SuppressExceptions` parameter, the whole point of `ExceptionOr<>` is to remove the need for this type of parameter.

rwlbuis replied:
I think I see what you mean, the spec has throwExceptions for this, will try to redo this part.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1344,7 +1344,7 @@ ExceptionOr<void> HTMLElement::showPopover()
 
         if (popoverState() != originalState)
-            return { };
+            return Exception { InvalidStateError, "The value of the popover attribute was changed while hiding the popover."_s };
```
I don't see this exception in the spec: https://html.spec.whatwg.org/#show-popover (Step 7.5)

@annevk Should this be added to the spec (in which case we should file an issue) or should the WPT be updated to match the spec?

nt1m replied:
cc @josepharhar @mfreed7 in case they have opinions about this particular bit.

nt1m replied:
(This exception seems OK to add IMO, though I just want to make sure everything gets put in the spec)

rwlbuis replied:
I did read 7.5 too quickly. I think things depend on the behaviour of throwExceptions being true and https://html.spec.whatwg.org/#check-popover-validity throwing an Exception. If the expectation is that the show popover algorithm passes that exception on (for throwExceptions being true) then everything is covered.

rwlbuis replied:
Anne pointed to this which I think resolves this discussion:
https://infra.spec.whatwg.org/#algorithm-control-flow

nt1m replied:
The way I read the algorithm doesn't require throwing the exception, since it's an "or", so we should be able to bail early if one of the conditions matches. So the popover validity check can be separate (which is how it currently is before this change).

Now if the "or" requires to evaluate both conditions, then we should probably move the checkPopoverValidity check up before the `if (popoverState() != originalState)` check. That would require no spec change.

annevk replied:
@nt1m I think that's correct. What I was clarifying to @rwlbuis is that algorithms rethrow, but in this case you don't hit the algorithm that throws.

rwlbuis replied:
I filed https://github.com/whatwg/html/issues/9037 for this.

rwlbuis replied:
I am splitting this part off (https://bugs.webkit.org/show_bug.cgi?id=254268) in anticipation of 9037 being resolved.

josepharhar replied:
Yeah I'm not sure what the motivation was for only checking for popover attribute state changes (between manual and auto) in show popover and not hide popover. It may have been intentional though

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1457,8 +1461,12 @@ void HTMLElement::popoverAttributeChanged(const AtomString& value)
+        hidePopoverInternal(FocusPreviousElement::Yes, FireEvents::Yes, SuppressExceptions::Yes);
+        if (!isConnected() || !popoverData() || oldValue != attributeWithoutSynchronization(popoverAttr))
+            return;
+    }
```
This doesn't match the spec either. I have a slight preference for what the spec does of not firing the events (which as the results means the type cannot be changed at this step). Can we update the WPT? cc @annevk 

https://html.spec.whatwg.org/#the-popover-attribute

<img width="914" alt="image" src="https://user-images.githubusercontent.com/1795135/224180595-8bc64304-729f-4f83-8a8c-a76505e7dcf8.png">

rwlbuis replied:
If you look at the pasted snippet, the spec seems to say focusPreviousElement=true, fireEvents=true and throwExceptions=false when calling "hide popover algorithm", so it does seem to match the current spec?

annevk replied:
Yeah, I think you're correct. I don't really like attribute changes firing events synchronously, but isn't the end of the world either I suppose. Chromium follows the specification I assume?

rwlbuis replied:
Yes, chromium follows the spec here, i.e. fires the event.

nt1m commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -52,6 +52,7 @@ enum class PopoverState : uint8_t {
     Auto,
     Manual,
 };
+enum class ThrowExceptions : bool { No, Yes };
```
As mentioned before, this enum shouldn't get introduced

rwlbuis replied:
Are you sure? The specification mentions throwExceptions as a parameter, similar to fireEvents:
https://html.spec.whatwg.org/multipage/popover.html#hide-popover-algorithm

Also we prefer enums to booleans.

nt1m replied:
`ExceptionOr<>` is precisely designed to avoid the need for this type of argument, it is functionally equivalent and should still keep us compliant.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1212,9 +1212,9 @@ ExceptionOr<Ref<ElementInternals>> HTMLElement::attachInternals()
+static ExceptionOr<void> checkPopoverValidity(HTMLElement& element, PopoverVisibilityState expectedState)
 {
-    if (!element.hasAttributeWithoutSynchronization(HTMLNames::popoverAttr))
+    if (!element.popoverData() || element.popoverState() == PopoverState::None)
```
The popoverState() check should be enough, see the definition: https://searchfox.org/wubkat/source/Source/WebCore/html/HTMLElement.cpp#1482

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1457,8 +1457,12 @@ void HTMLElement::popoverAttributeChanged(const AtomString& value)
+    if (oldPopoverState != PopoverState::None) {
+        auto oldValue = attributeWithoutSynchronization(popoverAttr);
+        hidePopoverInternal(FocusPreviousElement::Yes, FireEvents::Yes);
+        if (!isConnected() || !popoverData() || oldValue != attributeWithoutSynchronization(popoverAttr))
```
Why do we skip `ensurePopoverData().setPopoverState(newPopoverState);` in the case of `!isConnected()` ?

Does this work properly?
```
div = document.createElement("div"); // create disconnected div
div.popover = "auto";
div.popover = "manual";
document.body.append(div);
```

You should get the manual popover state after the two popover assignments.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1457,8 +1457,12 @@ void HTMLElement::popoverAttributeChanged(const AtomString& value)
+    if (popoverData() && popoverData()->visibilityState() == PopoverVisibilityState::Showing) {
+        auto oldValue = attributeWithoutSynchronization(popoverAttr);
+        hidePopoverInternal(FocusPreviousElement::Yes, FireEvents::Yes);
+        if (!isConnected() || !popoverData() || oldValue != attributeWithoutSynchronization(popoverAttr))
```
Checking for the popover showing state addresses the previous case, but the `!isConnected()` check would still break for this case (in fact Chrome has this bug):

```
var div = document.createElement("div");
document.body.append(div);
div.popover = "manual";

div.showPopover();

div.addEventListener("beforetoggle", (e) => {
  if (e.newState == "closed") {
    div.remove();
    setTimeout(() => {
      document.body.append(div);
      div.showPopover();
    }, 0);
  }
}, {once: true});

div.popover = "auto";
```
with this case, the resulting popover has manual behavior despite the attribute value being auto.

I wonder if there are similar bugs with the other two checks.

cc @josepharhar @mfreed7 heads up :)

mfreed7 replied:
Ohh, yeah that's not good. I'm open to suggestions on how to fix. I'd be supportive of adding a WPT that exposes this bug, also. @josepharhar any thoughts?

nt1m replied:
@mfreed7 The easiest fix is just to remove all these checks and stop firing events, like WebKit already currently does. I guess the "stop firing events" part is up for discussion, but I personally think it's reasonable in this case.

nt1m replied:
(Regarding WPTs, I'm out next week, so I'd appreciate if one of you can write one :) )

nt1m replied:
I filed this HTML spec issue 2 weeks ago fwiw: https://github.com/whatwg/html/issues/9034

rwlbuis committed:
[popover] Fix popover-attribute-basic.html
https://bugs.webkit.org/show_bug.cgi?id=253673

Reviewed by Tim Nguyen.

In popoverAttributeChanged when hiding the popover make sure we fire beforetoggle event.
Since this event may end up changing the popover attribute again, recompute the type
after hiding the popover.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-change-type-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-change-type.html: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-change-type-expected.txt: Added.
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):
(WebCore::HTMLElement::popoverAttributeChanged):

Canonical link: https://commits.webkit.org/263188@main
